fix(@aws-amplify/datastore): strictly define `null` vs `undefined` behavior on models

svidgen opened the pull request:
<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#pull-requests
-->

#### Description of changes
<!--
Thank you for your Pull Request! Please provide a description above and review
the requirements below.
-->

DataStore currently treats nullable model fields different before/after synchronization, which can lead to unexpected results when operating against or querying these instances. For example:

Given:

```graphql
type Model @model {
  id: ID!
  value: String!
  otherValue: String
}
```

If we save an instance:

```js
const saved = await DataStore.save(new Model({
    value: 'some value'
    /* otherValue: 'oh nevermind!' */
}));
```

We'll have this returned:

```json
{
  "value": "some value",
  "id": "a9f6e211-1d91-45c6-8fbe-46173fb9f09b"
}
```

If we query by ID *before* the instance is synchronized to the cloud, we see the same thing. And, if we query for models where `m.otherValue.eq(null)`, the model is *not* in the result set, because the field is actually `undefined`.

However, after synchronization, the the cloud storage provider (e.g., AppSync) will set optional fields to `null`:

```json
{
  "id": "a9f6e211-1d91-45c6-8fbe-46173fb9f09b",
  "value": "unique instance 0.21662066323138118",

  // LOOK!
  "otherValue": null, 

  "createdAt": "2023-01-27T18:02:40.093Z",
  "updatedAt": "2023-01-27T18:02:40.093Z",
  "owner": "test",
  "_version": 1,
  "_lastChangedAt": 1674842560119,
  "_deleted": null
}
```

The model will then appear in the result set if we search for `m => m.otherValue.eq(null)`.

This is an unintended behavior that customers should not have to work around. This PR normalizes that behavior so that optional fields default to `null` when unspecified or specified as `undefined`. (`undefined` can't actually be persisted in cloud storage.)

#### Issue #, if available
<!-- Also, please reference any associated PRs for documentation updates. -->



#### Description of how you validated changes

1. Added and updated unit tests.
2. Verified in sample apps.

#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

svidgen commented on packages/datastore/src/predicates/next.ts:
```diff
@@ -558,17 +592,41 @@ export class GroupCondition {
+			const predicateB =
+				this.withFieldConditionsOnly(negateChildren).toStoragePredicate();
+
+			// console.log(
+			// 	JSON.stringify(
+			// 		{
+			// 			predicateA: FlatModelPredicateCreator.getPredicates(predicateA!),
+			// 			predicateB: FlatModelPredicateCreator.getPredicates(predicateB!),
+			// 		},
+			// 		null,
+			// 		2
+			// 	)
+			// );
+
+			// console.log(
+			// 	JSON.stringify(
+			// 		{
+			// 			a: this,
+			// 			// b: this.toAST(),
+			// 			c: this.withFieldConditionsOnly(negateChildren),
+			// 			d: this.withFieldConditionsOnly(negateChildren).toAST(),
+			// 		},
+			// 		null,
+			// 		2
+			// 	)
+			// );
```
Pending cleanup.

svidgen commented on packages/datastore/src/predicates/next.ts:
```diff
@@ -558,17 +592,41 @@ export class GroupCondition {
 		// if conditions is empty at this point, child predicates found no matches.
 		// i.e., we can stop looking and return empty.
 		if (conditions.length > 0) {
-			const predicate = FlatModelPredicateCreator.createFromExisting(
+			const predicateA = FlatModelPredicateCreator.createFromExisting(
 				this.model.schema,
 				p =>
 					p[operator](c =>
 						applyConditionsToV1Predicate(c, conditions, negateChildren)
 					)
 			);
 
-			resultGroups.push(
-				await storage.query(this.model.builder, predicate as any)
-			);
+			const predicateB =
```
pending cleanup.

svidgen commented on packages/datastore/src/predicates/next.ts:
```diff
@@ -359,6 +376,24 @@ export class GroupCondition {
 		return [copied, extractedCopy];
 	}
 
+	withFieldConditionsOnly(negate: boolean) {
```
TODO: docstring.

svidgen commented on packages/datastore/src/predicates/next.ts:
```diff
@@ -169,6 +171,21 @@ export class FieldCondition {
 		};
 	}
 
+	negated(model: ModelMeta<any>) {
```
TODO: docstring.

svidgen commented on packages/datastore/__tests__/indexeddb.test.ts:
```diff
@@ -600,6 +608,8 @@ describe('Indexed db storage test', () => {
 			}
 		);
 
+		console.log({ sortedPersons });
```
TODO: remove.

svidgen commented on packages/datastore/__tests__/indexeddb.test.ts:
```diff
@@ -586,6 +586,14 @@ describe('Indexed db storage test', () => {
 		await DataStore.save(p2);
 		await DataStore.save(p3);
 
+		// TEMP
+		await DataStore.save(
+			new Person({
+				firstName: 'Decoy first',
+				lastName: 'Decoy last',
+			})
+		);
```
Remove.

svidgen commented on packages/datastore/src/storage/adapter/IndexedDBAdapter.ts:
```diff
@@ -495,7 +506,16 @@ class IndexedDBAdapter implements Adapter {
+					p.field === key &&
+					p.operator === 'eq' &&
+					// it's only valid if it's not nullish.
+					// (IDB will throw a fit if it's nullist.)
```
Typo. "nullist" -> "nullish"

svidgen commented on packages/datastore/src/storage/adapter/IndexedDBAdapter.ts:
```diff
@@ -443,6 +443,17 @@ class IndexedDBAdapter implements Adapter {
 
 		const records: T[] = (await (async () => {
+			//
+			// NOTE: @wirej explored removing this and letting query() take care of automatic
```
@wirej -> @svidgen

svidgen commented:
Converting back to draft for now. Tests against the sample app show that `readonly` fields are being included as inputs to the model input for create mutations. This wasn't caught by any tests. Adding tests and fixing ...

iartemiev commented:
From reading the code it seems like this doesn't address the scenario where I set a value to `undefined` in an update. Is that correct? 

For example:
Save new record, set optional value
```js
const saved = await DataStore.save(new Model({
  value: 'some value'
  otherValue: 'optional value'
}));
```
Update the optional value to `undefined`
```js
const updated = await DataStore.save(Model.copyOf(saved, (updated) => {
  updated.otherValue = undefined;
});

console.log(updated); // {  ..., otherValue: undefined }

// wait for sync
console.log(await DataStore.query(Model, saved.id)); // {  ..., otherValue: null }
```

svidgen commented:
> From reading the code it seems like this doesn't address the scenario where I set a value to `undefined` in an update. Is that correct?

Digging on this now. There was definitely a gap in test-cases for `copyOf()`. So, I'm filling those cases in, and I'll see about the update behavior. I'd forgotten we may have had existing bugs around that as well. But, it seems closely-related enough to take a look.

svidgen commented:
Putting this back into a draft state while I investigate your question @iartemiev. I'm seeing some odd behavior trying to add tests for it, and I want to rule out the possibility of introducing a regression on updates. 😓 

Thanks for poking at it!

svidgen commented:
@iartemiev I've added test cases to cover your question. Everything is working as expected -- I had reverted to a draft temporarily here out of an abundance of caution because I was seeing something unexpected in a test. But, it was a mistake in the test.

So, to answer your question, that case [*is* covered](https://github.com/aws-amplify/amplify-js/pull/10915/files#diff-10866b5ac7d469b38dd9928d9d7056a3a402904fef591ea8cfe773b4d75d63efR880). And, now it's tested as well. 😅  (See [DataStore.ts:2733](https://github.com/aws-amplify/amplify-js/pull/10915/files#diff-37f5c3d40431d210a7024204c661c85211ef5d7bb578d7a823a86375f75fcd18R2733-R2754) and [connectivityHandling.test.ts:134](https://github.com/aws-amplify/amplify-js/pull/10915/files#diff-68fad1b028a70ed30e99e616e38d45ec29a2993942226f28ff94cdb241bb3b25R134-R161).)

david-mcafee commented on packages/datastore/src/datastore/datastore.ts:
```diff
@@ -741,20 +740,56 @@ const castInstanceType = (
+	draft: Draft<T>
+) => {
+	for (const k of Object.keys(modelDefinition.fields)) {
+		if (draft[k] === undefined) (<any>draft)[k] = null;
```
Should we include some kind of warning, here? For instance, if the customer is explicitly setting this to `undefined`, and thus would be expecting it to be `undefined`?

svidgen replied:
I could see cases where this might be helpful. But, it'll also be *very* noisy, and I feel that it normalizes using warnings to tell the customer "normal things are happening" instead of "things you might not have meant to do!" Any time a customer saves an object and omits *any optional fields*, for example, they'd see the warning.

I still feel like I can be convinced this is a good idea. But, it feels like a hard sell at the moment.

dpilch replied:
I agree warning would be too noisy. I think docs update would be sufficient.

svidgen replied:
Sounds good. I've added a sprint item for the docs update.

david-mcafee commented on packages/datastore/src/storage/adapter/IndexedDBAdapter.ts:
```diff
@@ -495,7 +506,16 @@ class IndexedDBAdapter implements Adapter {
+					// it's only valid if it's not nullish.
+					// (IDB will throw a fit if it's nullish.)
+					p.operand !== null &&
+					p.operand !== undefined
```
What about the Async Adapter?

svidgen replied:
The error was specifically an indexed DB error. The largest barrage of failures that led to this change were the related model tests. And those tests are also executed in the Async Adapter suite. So, there's no Async Adapter impact that I could see.

david-mcafee commented on packages/datastore/src/predicates/next.ts:
```diff
@@ -141,7 +143,7 @@ export class FieldCondition {
 		public field: string,
 		public operator: string,
-		public operands: string[]
+		public operands: any[]
```
Why `any`?

svidgen replied:
That's really odd. I have literally no recollection of why I changed this.

Good catch!

david-mcafee reviewed (commented):
Left a few comments, mostly questions :)

iartemiev commented on packages/datastore/__tests__/DataStore.ts:
```diff
@@ -2699,15 +2699,58 @@ describe('DataStore tests', () => {
+			});
+
+			test('copyOf() setting optional field to null', () => {
+				const original = new Model({
+					field1: 'someField',
+					dateCreated: new Date().toISOString(),
+					optionalField1: 'defined value',
+				});
+				const copied = Model.copyOf(original, d => (d.optionalField1 = null));
+				expect(copied.optionalField1).toBe(null);
```
I already tested this locally, so I know it works as expected, but I think we should also ensure that  optional fields with a value that are omitted from the `copyOf` aren't accidentally nullified.

```suggestion
				const emailsVal = ['test@test.test'];
				const original = new Model({
					field1: 'someField',
					dateCreated: new Date().toISOString(),
					optionalField1: 'defined value',
					emails: emailsVal,
				});
				const copied = Model.copyOf(original, d => (d.optionalField1 = null));
				expect(copied.optionalField1).toBe(null);
				expect(copied.emails).toEqual(emailsVal);
```

iartemiev reviewed (dismissed):
LGTM 👍🏻
